sympy/polys/polytools.py

```
@@ -5307,6 +5314,120 @@ def half_gcdex(f, g, *gens, **args):
         return s, h
 
 
+@public
+def gcdex_steps(f, g):
```

I'm wondering if this should be made consistent with `gcdex` e.g.:
```python
In [3]: gcdex(x**2, x, x)
Out[3]: (0, 1, x)

In [4]: gcdex(Poly(x)**2, Poly(x))
Out[4]: (Poly(0, x, domain='QQ'), Poly(1, x, domain='QQ'), Poly(x, x, domain='QQ'))
```

---

I think thats a good idea. However, I'm not entirely sure what is going on in the body of `gcdex` (sympy.polys.polytools line 5442)
```python
def gcdex(f, g, *gens, **args):
    options.allowed_flags(args, ['auto', 'polys'])

    try:
        (F, G), opt = parallel_poly_from_expr((f, g), *gens, **args)
    except PolificationFailed as exc:
        domain, (a, b) = construct_domain(exc.exprs)

        try:
            s, t, h = domain.gcdex(a, b)
        except NotImplementedError:
            raise ComputationFailed('gcdex', 2, exc)
        else:
            return domain.to_sympy(s), domain.to_sympy(t), domain.to_sympy(h)

    s, t, h = F.gcdex(G, auto=opt.auto)

    if not opt.polys:
        return s.as_expr(), t.as_expr(), h.as_expr()
    else:
        return s, t, h
```
If the 'Polyfication' succeeds, then I think I understand whats going on:
- First, the arguments `f` and `g` are converted to polynomials, and `opt` stores whether any conversion waas actially needed (i.e. if `f` and `g` were already of type `Poly`, then `opt.polys = True`). 
- Then, the member method `gcdex` is called to do the extended euclidean algorithm. 
- Finally, the function returns the result as `Poly` if the inputs were of type `Poly`, and as `Expr` if the inputs were not of type `Poly`.

However, I'm not sure what happens in the case where polyfication fails. I dont believe that `gcdex_steps` is implemented for any of the specific domains, so I assume that this would always run into a `NotImplementedError`? 

For consistency with `gcdex` and `half_gcdex`, should `gcdex_steps` be a member method of `Poly`?

---

When polification fails a routine for domain elements should be used. It does not have to be a method on every domain but just a function that works with domain elements rather than Poly. I'm not sure what domains it makes sense for but this is so that you can do e.g.:
```python
In [1]: gcdex(2, 4)
Out[1]: (1, 0, 2)

In [2]: gcdex(2*x, 4)
Out[2]: (0, 1/4, 1)
```
I would probably just make a function that assumes that it can use divmod with the domain elements and implements the extended Euclidean algorithm.

---

When I run this with the current implementation (no fallback to working with domain elements) we get 
```python
for s in gcdex_steps(2*x, 4):
    print(s)
```
```
(0, 1, 4).
```
Is it problematic that the elements are not divided by `4`?

---

What domain does `gcdex_steps` use in this case? I think that `gcdex` always ensures that it is a polynomial with a field as ground domain.

---

In this case, I believe that it will be a polynomial with the integers as the ground domain. As it is currently, the `gcdex_steps` attempts to convert the arguments to polynomials and throws an error if it can't, rather than falling back on an alternative. 

Why is it that `gcdex` ensures this before doing the calculation? Is it to guarantee that it is working with a euclidean domain?

---

> Is it to guarantee that it is working with a euclidean domain?

I assume so. Is the extended Euclidean algorithm defined for univariate polynomials with integer coefficients?

---

$\mathbb{Z}[x]$ isn't a euclidean domain, so technically the answer is a no. [This stack exchange](https://math.stackexchange.com/questions/1300106/why-is-mathbbzx-not-a-euclidean-domain-what-goes-wrong-with-the-degree-f) link has some discussion about it.

I think the safest way is probably to just implement `gcdex_steps` in the same way as `gcdex`.

---

I agree.

---

It is the example I showed above: `gcdex(2, 4)`. Constructing polynomials fails if there are no symbols. Note:
```python
In [1]: gcdex(2, 4)
Out[1]: (1, 0, 2)

In [2]: gcdex(2, 4, x)
Out[2]: (0, 1/4, 1)
```
The first is gcdex in the domain `ZZ` and the latter is in `QQ[x]`. This one is `ZZ_I`:
```python
In [4]: gcdex(1+I, 2)
Out[4]: (-1, 1 + ⅈ, 1 + ⅈ)
```

---

My mistake, I thought I had checked that case. So am I right in assuming that the `PolyficationFailed` except block is intended for dealing with the cases you have given? That is, when the arguments to `gcdex` aren't polynomials, but are elements of some domain?

I've re-structured `gcdex_steps` to be similar to `gcdex`. However, I haven't written any implementations of `gcdex_steps` for particular domains, so the `PolyficationFailed` branch will always raise `ComputationFailed` at the moment.

---

Yes, the PolificationFailed block is for when the expressions can be represented in a domain but are not polynomials. These interfaces are a little strange but that is how it works. Ideally polynomials would just be an example of a domain rather than something separate.